### ISSUE TITLE
can't find reference nested in template

### ISSUE THREAD

**Issue by wrvsrx** [clangd/clangd#399]
I'm not sure if it's a bug or it's because clangd can't do that.

1. write a test.cpp:
```cpp
template<class T> void a(T input) {}
template<class T> void b(T input) {a(input);}
void c(int input) {a(input);}
```
2. call CocActionAsync('jumpReferences') when cursor is on "a(T input)".

It only shows reference in function "c".
![image](https://user-images.githubusercontent.com/42770726/82514359-2e012900-9b48-11ea-8d94-8eb0d077707f.png)

**Comment by HighCommander4** [comment]
This is similar to #258, except there is no instantiation of `b()` to find a concrete reference in, so we'd have to use some heuristics to find the reference in the uninstantiated template body.

**Comment by sam-mccall** [comment]
This is a bug, since go-to-definition works here, and isn't using heuristics.

The relevant AST node is a `UnresolvedLookupExpr` and this is handled (as the base `OverloadExpr`) by `targetDecl()`.
It *has* found the relevant `a` decl, but is considered unresolved because it's ADL-pending (e.g. if `ns::a(ns::Foo)` exists then it must be called when `b<ns::Foo>()` is instantiated).

However I think it's totally reasonable for find-refs to follow go-to-def here:
 - this is a reference to all visible overloads of `a`
 - with no instantiation, this is not a reference to `ns::a`
 - with an instantiation of `b<ns::Foo>()` it's fine to either report or not report the reference, being completely consistent here probably isn't feasible

The cause of the discrepancy is the use of `libIndex` rather than `FindTargets` for indexing references, I think. So we can fix this in libIndex or defer it until we switch.

**Comment by HighCommander4** [comment]
> This is a bug, since go-to-definition works here, and isn't using heuristics.

Just to clarify: you mean it's not using a _textual_ heuristic, right?

Assuming that an `OverloadExpr` resolves to one of its phase-one targets before instantiation is still a heuristic in my mind (albeit an AST-based one), much like performing name lookup in the primary template for a `DependentScopeDeclRefExpr`.

**Comment by HighCommander4** [comment]
This has been fixed since clangd 13. It was fixed by #675.
